### DISCUSSION TITLE
Is there a way to make drawing of many objects faster in avalonia (Skia)

### DISCUSSION THREAD

**Discussion by DrDryg** [AvaloniaUI/Avalonia#13149] [General]
Ive been experimenting drawing many small objects to a canvas and I'm very surprised by how slow it is. I want to be able to zoom and pan smoothly, but with just 50000 objects (small circles) it gets pretty slow (perhaps 5-10 fps) when panning. Im new UI development but considering how complex 3D worlds that can be rendered smoothly it feels like Im must be doing something wrong.  

Ive been digging into the code and it seems like for every change there is a redraw issued from the CPU. So drawing operations are performed on CPU and then things must be sent to GPU. I had expected that transforms like panning and zooming would be performed on GPU but that does not seem to be the case. 

So my questions are: Am I doing something wrong or is this the performance to be expected? Is there something I can do to make drawing of many primitive objects faster? (Im already drawing directly using SkCanvas.) What is actually GPU accelerated with Skia, and is that feature always on?

I have a ok laptop with a gtx 3070 card.

**Comment by timunie** [comment]
yeah rendering 5.000 objects is a heavy lifting, as you need to calculate arrange and measurement and other expensive stuff. So I hightly suggest to consider drawing on your own. You can for example use a RenderTargetBitmap or inherit from Control and override OnRender method. 

Good luck 🍀

**Reply by DrDryg** [reply]
Im already overriding the Render method and create a ICustomDrawOperation that draws directly to Skia. So there is no arrange or measurements going on. I simply draw many objects to a Skia Canvas. At first I suspected it to be what you mention that took a lot of time. But it seems to be the actual drawing operations. I had expected

**Reply by mgnslndh** [reply]
A few things I have noticed gave us better performance:

- Prefer SKPath over many calls to DrawLine, etc. Even though you might have many entities in your model that needs to be drawn you can still combine them into fewer amount of paths.
- SKPicture can be used to cache drawing operations.

Performance is still not good enough so hopefully we'll come up with more techniques to speed things up.

**Reply by timunie** [reply]
> Performance is still not good enough so hopefully we'll come up with more techniques to speed things up.

I wonder if SkiaSharp 3 will help with this a bit. Not yet ready to be shipped with Avalonia, but we have a PR to test at least. #12604

**Reply by DrDryg** [reply]
Thanks for the input! I will look into the use of Skia directly and look forward to a more performant version. What would be really nice as I understand how it works is if the amount of copying from cpu to gpu could be reduced. Moving, panning and zooming seems like operations you dont want to do on cpu.

**Reply by mgnslndh** [reply]
ZoomBorder is nice to use but I guess it mostly deals with input and converts that into a matrix for transformation of the child control. That works out of the box in many cases. However, when dealing with infinite canvases, custom drawing and lots of items in the model it might be better with control targeted for that use case. The control should work similar to ZoomBorder and handle input and transformation but instead of scaling the child it should calculate the viewport and delegate drawing of the viewport to something else (ICustomDrawOperation?). Performant panning could be achieved by rendering a bitmap that is larger than the viewport and just moving that as the viewport is panned. For zooming I guess you wouldn't want to delegate during zooming but only when the zoom has finished. The control should take care of zooming in and out quickly and smoothly :) Basically, make it easier to get good out of the box performance for panning and zooming. Just some thoughts. It could be that ZoomBorder can be used but we haven't cracked the "how" yet :)

**Comment by sn4k3** [comment]
How often do you need to redraw? If not too often you can draw to a Bitmap. I have made a control that Zooms and Pans on a bitmap with fast performance, if you want to try: [AdvancedImageBox](https://www.nuget.org/packages/UVtools.AvaloniaControls#readme-body-tab) / [Video-demo](https://www.youtube.com/watch?v=bIr6P4dDlHc), in minute 1:15 you can see how fast user can draw 1700 squares in live image, each square trigger a render to UI, if you draw them all at once and trigger the render only at end it would be almost instant.
Also if you can draw all objects in one pass would be ideal, not sure if SK can do it.

**Reply by mgnslndh** [reply]
Nice app! I tried UVTools and the solutions looks interesting but I will need to take a closer look to know if I can use this approach. I will probably have some questions for you :)

**Reply by DrDryg** [reply]
Working with bitmaps certainly helped. It is fast to pan in a single bitmap. But redrawing many objects live is still "slow". More than a few 1000 rects or similar makes the frame-rate drop to 10-20Hz. So in a sense a few ten 10ms is almost instant except that the programs appears lagging. Pre-drawing the many (for my complex items) to small images and drawing these to a large image helps a lot. But one has to resort to Skia Images. Using Bitmaps is very slow since they are on the CPU. These are my experiences at least. Take them with a grain of salt since I might not do everything in the most optimal way.

**Reply by sn4k3** [reply]
In my app I use openCV to draw, that means from openCV Mat i have to convert to Bitmap, image sample (QHD):

![image](https://github.com/AvaloniaUI/Avalonia/assets/113281/bf7b6cb7-5ee0-407c-bb5c-368334213683)
In debug mode, takes from 6ms to 11ms to: Decode, convert to bitmap, draw to UI 

Now if I add all my processes:

![image](https://github.com/AvaloniaUI/Avalonia/assets/113281/fa60ba96-9f24-495a-a6ae-ede3a7be4774)
In debug mode, takes from 28ms to 31ms to: Decode, convert to bitmap, draw two bondary rectangles, draw all blobs rectangles, compute contours, draw enclosing circles, draw hollow areas, draw centroids, triagulate from pixels and draw the triangulation lines, draw text, compute edges, draw all edges and draw to UI.
It also can compare below and above layers, pixel by pixel, and draw differences over the image and only takes an additional of 8ms to 10ms for such hard task.
Given the amount of processing is doing, is amazing the short time it takes but openCV library it's optimized and big, so far the fastest I manage to find to work with images and drawings.
If you notice that sample have 9382 triangles, on my app that means a foreach in every triangle and draw a single line per each vertice, which ends in 9382 x 3 lines draw calls = 28,146 lines plus 28,146 circles to mark each vertice! which is amazing :) Oh and contours are made from tons of points which are also draw from line calls connecting all points... So the workload is huge here. If I had all calculations in cache the draw methods of openCV are almost instant, the most time it takes is not from draw but from calculations it perform to find stuff.

With all that processing active see how fast it render and navigate 1000 images, note that every time it switches image all calculations are perfomed again (They are not cached, only image is cached under LZ4 which require to decode to mat and to bitmap)

https://github.com/AvaloniaUI/Avalonia/assets/113281/1abbcfa3-f63d-44b9-add9-e97c1ee94f51

Also look into code and realize how this monster eats code, it's sick and I amaze myself how fast it is given the ton of code and processing in a draw to UI :) https://github.com/sn4k3/UVtools/blob/fe8bf8ff8403c1c790c04fd58e593b3c19459a62/UVtools.UI/MainWindow.LayerPreview.cs#L862-L1514

I believe using openCV you can optimize your draw using contours to draw all at once with a single call, however using such library may not be optimal for your project. ImageSharp is not close to the speed of openCV but you can also try it.

**Reply by DrDryg** [reply]
Very interesting and thanks for you long and detailed answer.
